[#16356] Add support for resource manager tags for google_compute_instance_template

LucaPrete opened the pull request:
Adds support for resource manager tags for google_compute_instance_template.

fixes https://github.com/hashicorp/terraform-provider-google/issues/16356
fixes https://github.com/hashicorp/terraform-provider-google/issues/15903
fixes https://github.com/hashicorp/terraform-provider-google/issues/12952

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: added `resource_manager_tags` and `disk.resource_manager_tags` for `google_compute_instance_template`
```

```release-note:enhancement
compute: added `resource_manager_tags` and `disk.resource_manager_tags` for `google_compute_region_instance_template`
```

modular-magician commented:
Hello! I am a robot. It looks like you are a: ~Community Contributor~ Googler ~Core Contributor~. Tests will run automatically.

@zli82016, a repository maintainer, has been assigned to review your changes. If you have not received review feedback within 2 business days, please leave a comment on this PR asking them to take a look.

You can help make sure that review is quick by [doing a self-review](https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/) and by [running impacted tests locally](https://googlecloudplatform.github.io/magic-modules/get-started/run-provider-tests/).

LucaPrete commented:
While I've checked manually that tags have been set, here and in `google_compute_instance` I'd like to check whether tags have been really created/updated.

In order to do so, I'll need to
1) list the instance tag key/values (returned with a different id from the one given in input)
2) call the resource manager v3 APIs ([seems I'd need to use a TagKeysClient for this?](https://pkg.go.dev/cloud.google.com/go/resourcemanager@v1.9.4/apiv3#TagKeysCallOptions)) to see whether it matches the key/value set

Do you think it makes sense to do it directly here or should we send this in first?

I'm not even sure it's a common practice calling another api (i.e. resourcemanager api from a compute test)

LucaPrete commented:
cc @rileykarson

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9612-old..auto-pr-9612) ( 2 files changed, 98 insertions(+), 5 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9612-old..auto-pr-9612) ( 2 files changed, 98 insertions(+), 5 deletions(-))

zli82016 commented:
> While I've checked manually that tags have been set, here and in `google_compute_instance` I'd like to check whether tags have been really created/updated.
> 
> In order to do so, I'll need to
> 
> 1. list the instance tag key/values (returned with a different id from the one given in input)
> 2. call the resource manager v3 APIs ([seems I'd need to use a TagKeysClient for this?](https://pkg.go.dev/cloud.google.com/go/resourcemanager@v1.9.4/apiv3#TagKeysCallOptions)) to see whether it matches the key/value set
> 
> Do you think it makes sense to do it directly here or should we send this in first?
> 
> I'm not even sure it's a common practice calling another api (i.e. resourcemanager api from a compute test)

Just want to confirm that do you want to check if the tags have been really created/updated in the test?

modular-magician commented:
$\textcolor{red}{\textsf{The provider crashed while running the VCR tests in REPLAYING mode}}$ 
$\textcolor{red}{\textsf{Please fix it to complete your PR}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9612/artifacts/69e4f7d4-c8ff-4793-8c52-97fdddf7167d/build-log/replaying_test.log)

LucaPrete commented:
> > While I've checked manually that tags have been set, here and in `google_compute_instance` I'd like to check whether tags have been really created/updated.
> 
> > 
> 
> > In order to do so, I'll need to
> 
> > 
> 
> > 1. list the instance tag key/values (returned with a different id from the one given in input)
> 
> > 2. call the resource manager v3 APIs ([seems I'd need to use a TagKeysClient for this?](https://pkg.go.dev/cloud.google.com/go/resourcemanager@v1.9.4/apiv3#TagKeysCallOptions)) to see whether it matches the key/value set
> 
> > 
> 
> > Do you think it makes sense to do it directly here or should we send this in first?
> 
> > 
> 
> > I'm not even sure it's a common practice calling another api (i.e. resourcemanager api from a compute test)
> 
> 
> 
> Just want to confirm that do you want to check if the tags have been really created/updated in the test?

@zli82016 yes.

LucaPrete commented:
/gcbrun

LucaPrete commented:
This seemed to be failing btw for nothing related to my code but to Dataproc:

```
=== NAME  TestAccDataprocJobIamPolicy
    testing_new.go:84: Error running post-test destroy, there may be dangling resources: exit status 1
        
        Error: Error setting IAM policy for Dataproc Job ci-test-project-188019/us-central1/tf-test-iam-c4kq37bwxh: Post "https://dataproc.googleapis.com/v1/projects/ci-test-project-188019/regions/us-central1/jobs/tf-test-iam-c4kq37bwxh:setIamPolicy?alt=json&prettyPrint=false": Requested interaction not found
        
--- FAIL: TestAccDataprocJobIamPolicy (65.80s)
```

Trying to rerun the tests to see if the error was fixed somewhere else...

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9612-old..auto-pr-9612) ( 2 files changed, 98 insertions(+), 5 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9612-old..auto-pr-9612) ( 2 files changed, 98 insertions(+), 5 deletions(-))

modular-magician commented:
$\textcolor{red}{\textsf{The provider crashed while running the VCR tests in REPLAYING mode}}$ 
$\textcolor{red}{\textsf{Please fix it to complete your PR}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9612/artifacts/696490de-9dba-4b32-aa3e-591a11479738/build-log/replaying_test.log)

zli82016 commented:
It looks like the error for crashed provider is related. Can you please try to fix it? Thanks.

I found this in the logs
```
panic: Invalid address to set: []string{"disk", "0", "resource_manager_tags"}

goroutine 195922 [running]:
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*ResourceData).Set(0xc023b74e80, {0x410d5e7, 0x4}, {0x3855ea0, 0xc0029ca018})
	/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.0/helper/schema/resource_data.go:230 +0x291
github.com/hashicorp/terraform-provider-google-beta/google-beta/services/compute.resourceComputeRegionInstanceTemplateRead(0xc023b74e80, {0x40e0d80?, 0xc01e0be000})
	/go/src/github.com/hashicorp/terraform-provider-google-beta/google-beta/services/compute/resource_compute_region_instance_template.go:1195 +0xd48
github.com/hashicorp/terraform-provider-google-beta/google-beta/services/compute.resourceComputeRegionInstanceTemplateCreate(0x0?, {0x40e0d80?, 0xc01e0be000})
	/go/src/github.com/hashicorp/terraform-provider-google-beta/google-beta/services/compute/resource_compute_region_instance_template.go:1086 +0xd92
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x49b0778?, {0x49b0778?, 0xc01898a540?}, 0xd?, {0x40e0d80?, 0xc01e0be000?})
	/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.0/helper/schema/resource.go:695 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc001bf20e0, {0x49b0778, 0xc01898a540}, 0xc00b0fe750, 0xc023b74d00, {0x40e0d80, 0xc01e0be000})
	/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.0/helper/schema/resource.go:837 +0xa85
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0137c5200, {0x49b0778?, 0xc01898a210?}, 0xc017c57900)
	/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.0/helper/schema/grpc_provider.go:1021 +0xe8d
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.muxServer.ApplyResourceChange({0xc019a58fc0, 0xc019a59020, {0xc0177d3920, 0x2, 0x2}, {0x0, 0x0, 0x0}, {0x0, 0x0, ...}, ...}, ...)
	/go/pkg/mod/github.com/hashicorp/terraform-plugin-mux@v0.8.0/tf5muxserver/mux_server_ApplyResourceChange.go:27 +0x102
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc01aa6ee60, {0x49b0778?, 0xc008238e40?}, 0xc00581c7e0)
	/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/tf5server/server.go:818 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x4049e20?, 0xc01aa6ee60}, {0x49b0778, 0xc008238e40}, 0xc023b74800, 0x0)
	/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc005f55860, {0x49b0778, 0xc008238d80}, {0x49ba778, 0xc0180bf520}, 0xc00d1e6c60, 0xc01e4f1950, 0x65b3600, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.59.0/server.go:1343 +0xe49
google.golang.org/grpc.(*Server).handleStream(0xc005f55860, {0x49ba778, 0xc0180bf520}, 0xc00d1e6c60)
	/go/pkg/mod/google.golang.org/grpc@v1.59.0/server.go:1737 +0xca6
google.golang.org/grpc.(*Server).serveStreams.func1.1()
	/go/pkg/mod/google.golang.org/grpc@v1.59.0/server.go:986 +0x8c
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/pkg/mod/google.golang.org/grpc@v1.59.0/server.go:997 +0x15c
FAIL	github.com/hashicorp/terraform-provider-google-beta/google-beta/services/compute	1647.405s
```

zli82016 reviewed (changes requested):
Need to fix provider crash

zli82016 commented:
The method `flattenDisks` is a shared function and also called in [resource_compute_region_instance_template.go.erb](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/compute/resource_compute_region_instance_template.go.erb#L1201). 

It looks like it needs support resource manager tags in `google_regional_compute instance_template`. API doc is here https://cloud.google.com/compute/docs/reference/rest/v1/regionInstanceTemplates

zli82016 commented on mmv1/third_party/terraform/services/compute/resource_compute_instance_template.go.erb:
```diff
@@ -1350,10 +1372,9 @@ func resourceComputeInstanceTemplateCreate(d *schema.ResourceData, meta interfac
 		ReservationAffinity:        reservationAffinity,
 	}
 
-	if _, ok := d.GetOk("effective_labels"); ok {
-		instanceProperties.Labels = tpgresource.ExpandEffectiveLabels(d)
```
`effective_labels` is introduced in Terraform Google Provider 5.0 and should stay here.

https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/version_5_upgrade#provider

zli82016 commented on mmv1/third_party/terraform/services/compute/resource_compute_instance_template_test.go.erb:
```diff
@@ -3674,6 +3699,51 @@ resource "google_compute_instance_template" "template" {
+    boot         = true
+
+    resource_manager_tags = {
+	  "tagKeys/${google_tags_tag_key.key.name}" = "tagValues/${google_tags_tag_value.value.name}"
```
indentation should be 2 spaces.

zli82016 commented:
Can you please add the new fields to `Example Usage` and `Argument Reference` in [the provider doc](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown)? Thanks.

LucaPrete commented:
@zli82016 it should all be fixed now. Fingers crossed the tests all pass now!

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9612-old..auto-pr-9612) ( 6 files changed, 203 insertions(+), 3 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9612-old..auto-pr-9612) ( 6 files changed, 203 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `858` 
Passed tests `790` 
Skipped tests: `65` 
Affected tests: `3` 

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>compute</li></ul></blockquote></details> 

#### Action taken 
<details> <summary>Found 3 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccComputeRegionInstanceTemplate_resourceManagerTags|TestAccComputeRegionPerInstanceConfig_removeInstanceOnDestroy|TestAccComputeInstanceTemplate_resourceManagerTags </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccComputeRegionInstanceTemplate_resourceManagerTags`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9612/artifacts/e51f1bb9-d44e-4709-9215-0ae2eeb21058/recording/TestAccComputeRegionInstanceTemplate_resourceManagerTags.log)]
`TestAccComputeRegionPerInstanceConfig_removeInstanceOnDestroy`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9612/artifacts/e51f1bb9-d44e-4709-9215-0ae2eeb21058/recording/TestAccComputeRegionPerInstanceConfig_removeInstanceOnDestroy.log)]
`TestAccComputeInstanceTemplate_resourceManagerTags`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9612/artifacts/e51f1bb9-d44e-4709-9215-0ae2eeb21058/recording/TestAccComputeInstanceTemplate_resourceManagerTags.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{green}{\textsf{All tests passed!}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9612/artifacts/e51f1bb9-d44e-4709-9215-0ae2eeb21058/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-9612/artifacts/e51f1bb9-d44e-4709-9215-0ae2eeb21058/recording) for each test

LucaPrete commented:
@shuyama1 I know @zli82016 is OOO. Do you know if there's anybody else we can add so we can finalize the review and possibly merge this?

zli82016 commented:
Thanks for making the changes. 

I updated the release notes and added the related github issues in the PR's description.

Please remember to do those for other PRs.

zli82016 reviewed (approved):
LGTM

LucaPrete commented:
Thank you!

Il giorno mar 2 gen 2024 alle 19:28 Zhenhua Li ***@***.***>
ha scritto:

> Thanks for making the changes.
>
> I updated the release notes and added the related github issues in the
> PR's description.
>
> Please remember to do those for other PRs.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/magic-modules/pull/9612#issuecomment-1874390533>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARY7UBB23YVLCFZVKIFAJDYMRGTZAVCNFSM6AAAAABANJBOUKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNZUGM4TANJTGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
